bugsnag-android-core/src/main/java/com/bugsnag/android/internal/JournaledDocument.kt

```diff
@@ -59,39 +83,57 @@ constructor(
      * Next, it will apply the command to the document, and then add it to the journal itself.
      */
     fun addCommand(command: Journal.Command) {
-        if (!isOpen) {
-            throw IllegalStateException("Cannot add commands to a closed document")
-        }
-        try {
-            command.serialize(journalStream)
-        } catch (ex: BufferOverflowException) {
-            snapshot()
-            command.serialize(journalStream)
+        synchronized(journal) {
+            if (!isOpen) {
+                throw IllegalStateException("Cannot add commands to a closed document")
+            }
+            try {
+                command.serialize(journalStream)
+            } catch (ex: BufferOverflowException) {
+                snapshot()
+                command.serialize(journalStream)
+            }
+            command.apply(document)
+            journal.add(command)
         }
-        command.apply(document)
-        journal.add(command)
     }
 
     /**
      * Save a current document snapshot and clear the journal
      */
     fun snapshot() {
-        if (!isOpen) {
-            throw IllegalStateException("Cannot snapshot a closed document")
+        synchronized(journal) {
+            snapshotInternal()
+        }
+    }
+
+    /**
+     * Save a current snapshot if the journal has passed the high water mark.
+     * Use this in a BG thread to reduce the incidence of snapshotting on the critical path.
+     */
+    fun snapshotIfHighWater() {
```

Comment:
It'd be worth considering whether every function in this class needs to be public - this one seems like it could be `private`?

Reply:
This function is designed to be called by a periodic housekeeping thread so that the actual act of snapshotting happens outside of the critical path as much as possible, so it needs to be public.

Reply:
I don't have a good understanding of why the function needs to be public if `highWater` is a constructor parameter that controls when the function is called, and the function is only called internally. Is there other usage of this function anticipated than what's present at the moment?

Reply:
This function is never called internally; it needs a controller object to call it.

The lines of responsibility are:
* `JournaledDocument` decides if it is in "high water" state (using the passed in constructor value) and snapshots if it is.
* `BugsnagMainJournal` decides when these "high water" checks occur.

`BugsnagMainJournal` will be triggering `document` (`JournaledDocument`) like this:

```kotlin
Thread {
    while(true) {
        Thread.sleep(highWaterPollingIntervalMS)
        document.snapshotIfHighWater()
    }
}
```

Reply:
Given that the snapshot & `addCommand` functions are both guarded by the `journal` monitor lock, I'm not clear on what a background thread adds here? Is it just so that snapshots can run concurrently with read operations?

Reply:
Yes, the locks protect against concurrent access (read and write) while the snapshot is going on.

The background thread is to push as many snapshot operations as possible onto that background thread so that they don't happen on the critical path. If the journal file overflows, it MUST be immediately flushed to disk to avoid data loss. So every time the high water code triggers on that separate thread, it mitigates an upcoming overflow, which means one less call to snapshot on the critical path.